Add apomorphy-based phyloref

gaurav opened the pull request:
This PR adds an apomorphy-based phyloreference for testing, "Testudinata".

Should be merged after PR #61, PR #66 and PR #71.

hlapp reviewed (changes requested):
So a UBERON term cannot be an exact match with an instance of a phenotype. This needs to be fixed. You could say that the phenotype
```
ro:inheres_in some obo:UBERON_0008271
```
(in OWL Manchester, not n-triples RDF, but you get the idea).

Second, we're extending the `testcase` ontology, which strikes me as a non-starter too. Either we do need an ontology/vocabulary property for this (in which case the Phyloreferencing Ontology was made for precisely this purpose), or we don't. What you say is conflicting on this question.

hlapp commented:
@balhoff or @wdahdul can you remind me how we encode a presence phenotype? `pato:present and inheres_in some <entity>`?

balhoff commented:
@hlapp yes that's correct. Although in the KB we wrap that in 'has part', like `has_part some (pato:present and inheres_in some <entity>)`. This is to align the model with what's found in the phenotype ontologies.

gaurav commented:
Thanks for the input, @balhoff!

Okay, so it sounds like I need to make three or four changes here:
1. Change `testcase:apomorphy` to `phyloref:apomorphy`, and open an issue on the Phyloref Ontology for this new term.
2. Change the JSON property field `exactMatch` to something like `phenotype`.
3. Change the value in the generated JSON-LD file so that it looks something like this:
```
  phyloref:apomorphy [ a owl:Class, <https://semanticscience.org/resource/SIO_010056> ;
    obo:IAO_0000115 "A 'complete turtle shell' is herein defined as a composite structure consisting of a carapace with interlocking costals, neurals, peripherals, and a nuchal, together with the plastron comprising interlocking epi-, hyo-, meso- (lost in Testudo graeca), hypo-, xiphiplastra and an entoplastron. These are articulated with one another along a bridge. Additional elements may be present as well." ;
    owl:intersectionOf (
      pato:present 
      [ a owl:Restriction
        owl:onProperty RO:0000052 # Inheres_in
        owl:someValuesFrom obo:UBERON_0008271
      ]
    )
  ] ;
```
4. Optionally, we could also go one step further: we could add a field to the apomorphy in JSON that indicates whether the phenotype should be present or absent in the JSON-LD (maybe `phenotypePresent` as a boolean field?), and then generate `pato:present` or `pato:absent` appropriately. What do you think, @hlapp?

hlapp commented:
* Note that `a owl:Class` is redundant. `<https://semanticscience.org/resource/SIO_010056>` is already a class.
* I don't think a `phenotypePresent` attribute in the JSON-LD makes much sense. But you could have a property for `quality` (or `phenotypicQuality`), which if provided would then need to be a subclass of `pato:quality`.

balhoff commented:
If I understand the context correctly, there are some issues with the generated RDF above. If `phyloref:apomorphy` is a class, then it cannot also be an instance of `<https://semanticscience.org/resource/SIO_010056>`, unless you are doing some kind of meta modeling using punning (if you are, I take it back 🙂). The subject of `owl:intersectionOf` must be a blank node for this RDF to be an OWL serialization. You would connect your named class to this blank node using `owl:equivalentClass`.

hlapp commented:
Good point @balhoff. I was saw _Change the value in the generated JSON-LD file_, which made me automatically assume that this wasn't going to be the generated OWL serialization yet. (That is generated from Phyx, which is a JSON-LD format.)

gaurav commented:
Thanks for the feedback, both of you! I've made some changes to the code, so that this [how the apomorphy now looks like in Phyx](https://github.com/phyloref/phyx.js/blob/e07f1b4022f70dd8c671bafcc2b71489d2907438/test/examples/correct/testudinata_phylonym.json):

```json
"apomorphy": {
    "phenotype": "http://purl.obolibrary.org/obo/UBERON_0008271",
    "phenotypicQuality": "http://purl.obolibrary.org/obo/PATO_0000467",
    "definition": "A 'complete turtle shell' is herein defined as a composite structure consisting of a carapace with interlocking costals, neurals, peripherals, and a nuchal, together with the plastron comprising interlocking epi-, hyo-, meso- (lost in Testudo graeca), hypo-, xiphiplastra and an entoplastron. These are articulated with one another along a bridge. Additional elements may be present as well."
}
```

Translated into Turtle/OWL, it now looks like this (as translated from the [N-Quads file](https://github.com/phyloref/phyx.js/blob/e07f1b4022f70dd8c671bafcc2b71489d2907438/test/examples/correct/testudinata_phylonym.nq)):
``` turtle
phyloref:apomorphy [ a owl:Class ;
    obo:IAO_0000115 "A 'complete turtle shell' is herein defined as a composite structure consisting of a carapace with interlocking costals, neurals, peripherals, and a nuchal, together with the plastron comprising interlocking epi-, hyo-, meso- (lost in Testudo graeca), hypo-, xiphiplastra and an entoplastron. These are articulated with one another along a bridge. Additional elements may be present as well." ;
    rdfs:subClassOf <https://semanticscience.org/resource/SIO_010056> ;
    owl:equivalentClass [ a owl:Class ;
        owl:intersectionOf ( obo:PATO_0000467 [ a owl:Restriction ;
            owl:onProperty obo:RO_0000052 ;
            owl:someValuesFrom obo:UBERON_0008271 ] ) ] ] ;
```

Does that look right? It doesn't quite work in Protege yet, but I think that's because the apomorphy itself isn't a named class -- it's currently an anonymous class. We could mandate that the user provide an `@id` for each apomorphy, but I don't think that's a good idea.

Alternatively, I could just shelve the idea of translating the apomorphy into any kind of logical expression for now. I think we've identified the three Phyx fields we would need to express a phenotype in OWL (`definition`, `phenotype` and `phenotypicQuality`). So I think I'll just modify this PR so that we no longer generate a logical expression at all for the apomorphy and call it good. What do you think, @hlapp?

balhoff commented:
> Does that look right? It doesn't quite work in Protege yet, but I think that's because the apomorphy itself isn't a named class -- it's currently an anonymous class. 

Right, you can't really make an anonymous class that isn't some kind of expression (like the nested intersection blank node). And you can't use a class expression as the subject of an annotation property. 😕

hlapp commented:
@gaurav note that in the Turtle/OWL both the `a owl:Class` statements are redundant (as they are implied). Furthermore, I think we need to either put a phenotype under the `phenotype` key in the JSON-LD, or rename the key. As is, it is confusing at best, and more likely misleading. If you wanted a value for the entity bearing the phenotype, then the key should be `entity`, not `phenotype`. Or name it `phenotypeBearer` if you wanted something more similar to `phenotypicQuality`.

As for generating the logical expression, I agree this should be out of scope for now. We could do so for a default case (presence of some anatomical entity; or even a little more generally, a single non-relational quality inhering in a  single entity), but as they go phenotypes can get very complex, and so can their logical expressions. Capturing that should be the subject of a future grant.

gaurav commented:
Hi everybody!

> Right, you can't really make an anonymous class that isn't some kind of expression (like the nested intersection blank node). And you can't use a class expression as the subject of an annotation property. 😕

Yeah. I guess we could have turned `phyloref:apomorphy` into an object property, but we'd still need to provide an identifier for the phenotype somehow. I'm now convinced this is out of scope for us.

> Furthermore, I think we need to either put a phenotype under the `phenotype` key in the JSON-LD, or rename the key. As is, it is confusing at best, and more likely misleading. If you wanted a value for the entity bearing the phenotype, then the key should be `entity`, not `phenotype`. Or name it `phenotypeBearer` if you wanted something more similar to `phenotypicQuality`.

I like `phenotypeBearer`! So now the JSON representation looks like this:

``` json
"apomorphy": {
    "phenotypeBearer": "http://purl.obolibrary.org/obo/UBERON_0008271",
    "phenotypicQuality": "http://purl.obolibrary.org/obo/PATO_0000467",
    "definition": "A 'complete turtle shell' is herein defined as a composite structure consisting of a carapace with interlocking costals, neurals, peripherals, and a nuchal, together with the plastron comprising interlocking epi-, hyo-, meso- (lost in Testudo graeca), hypo-, xiphiplastra and an entoplastron. These are articulated with one another along a bridge. Additional elements may be present as well."
}
```

I've also added a bit of logic so that the phenotypeQuality in the OWL output will default to [`obo:PATO_0000467`](http://purl.obolibrary.org/obo/PATO_0000467) ("present") if it is not present in the Phyx file. Does that sound like a good idea?

> As for generating the logical expression, I agree this should be out of scope for now. We could do so for a default case (presence of some anatomical entity; or even a little more generally, a single non-relational quality inhering in a single entity), but as they go phenotypes can get very complex, and so can their logical expressions. Capturing that should be the subject of a future grant.

Sounds good! I've taken that logic out. The OWL representation is pretty simple now:

``` json
phyloref:apomorphy [
    obo:IAO_0000115 "A 'complete turtle shell' is herein defined as ... elements may be present as well."
] ;
```

Unless we want to include `phenotypeBearer` and `phenotypicQuality` in the OWL representation, I think that's good enough for now!

hlapp commented:
> I've also added a bit of logic so that the phenotypeQuality in the OWL output will default to obo:PATO_0000467 ("present") if it is not present in the Phyx file. Does that sound like a good idea?

Sounds risky to me. What if the apomorphy is loss of a trait, and the curator forgets to change the default?

> The OWL representation is pretty simple now:
> 
> ```turtle
> phyloref:apomorphy [
>     obo:IAO_0000115 "A 'complete turtle shell' is herein defined as ... elements may be present as well."
> ] ;
> ```

Shouldn't we add a type (subclass of SIO_010056)?

gaurav commented:
Sounds good! I've fixed those in a663d94dbf03a8f47 and 5fe258a020f1ec9. I've updated the JSON and OWL representations in the description of this PR.

Note that I used `@type` (i.e. `rdf:type`), so it's not a *subclass* (`rdfs:subClassOf`) of SIO_010056, but rather an *instance* of SIO_010056. I tried using `subClassOf` but this causes an error in Protege (`Entity not properly recognized, missing triples in input?`), probably because -- as Jim mentioned previously -- an anonymous blank node can't be an OWL class. I tried various ways of convincing it that the anonymous blank node should be treated as an RDFS:class and not an OWL:class, but was unsuccessful.

hlapp commented on docs/context/development/schema.json:
```diff
@@ -332,6 +339,25 @@
+          },
+          "apomorphy": {
+            "type": "object",
+            "description": "For a phyloreference using an apomorphy, this field describes the character being used as the apomorphy",
```
```suggestion
            "description": "For a phyloreference using an apomorphy, this field describes the phenotype (or trait) being used as the apomorphy",
```

hlapp replied:
An apomorphy is not a character but a character state (or more generally, a phenotype or trait).

hlapp commented on docs/context/development/schema.json:
```diff
@@ -332,6 +339,25 @@
+            "additionalProperties": true,
+            "properties": {
+              "definition": {
+                "description": "A description of this apomorphy."
```
```suggestion
                "description": "A natural language description of this apomorphy."
```

hlapp replied:
We should be specific here whether that description needs to follow some standard or not.

hlapp commented on docs/context/development/schema.json:
```diff
@@ -332,6 +339,25 @@
+                "description": "The quality of this phenotype. Must be a subclass of pato:quality."
+              },
+              "phenotypeBearer": {
+                "description": "The phenotype being used as an apomorphy."
```
No, this is not the phenotype, it's the bearer of the phenotype. 😞

gaurav replied:
So, this should be "The bearer of the phenotype being used as an apomorphy."? I'm having a hard time wrapping my head around the idea that an UBERON term (such as "[turtle shell](http://purl.obolibrary.org/obo/UBERON_0008271)") is the bearer of the phenotype, whereas I would intuitively think of the bearer of the phenotype as an individual turtle with such as shell. However, if that is the right phrasing from an UBERON/PATO perspective (i.e. if that's what people who think in terms of phenotypeBearer/phenotypicQuality would understand), then that is absolutely what we should use.

hlapp replied:
The individual organism exhibits the phenotype. The turtle shell is the bearer of the phenotypic quality ("present") (and to be correct, not of the phenotype, which is the combination of quality and bearing entity; I corrected that).

hlapp replied:
So `phenotypeBearer` is therefore not quite correct, you are right. The entity is the bearer of the phenotypic quality, not the phenotype. Maybe use `bearingEntity` instead?

gaurav replied:
Sounds good! I've renamed that property to `bearingEntity` in fb09cc0263a.

hlapp commented on src/wrappers/PhyxWrapper.js:
```diff
@@ -270,12 +270,19 @@ class PhyxWrapper {
+    if (!has(jsonld, 'owl:imports')) jsonld['owl:imports'] = [];
+    if (!Array.isArray(jsonld['owl:imports'])) jsonld['owl:imports'] = [jsonld['owl:imports']];
+    jsonld['owl:imports'].push('http://raw.githubusercontent.com/phyloref/curation-workflow/develop/ontologies/phyloref_testcase.owl');
+    jsonld['owl:imports'].push('http://ontology.phyloref.org/2018-12-14/phyloref.owl');
```
Any reason we have to have the versioned imports rather than just the most current version?

gaurav replied:
The idea here is that this version of Phyx is designed to work with the 2018-12-14 version of the Phyloref.owl and TCAN.owl ontologies. If we make changes to these ontologies, I assume that we will need to make changes in Phyx.js to reflect this change (for instance, if some minor part of the behavior of `phyloref:excludes_lineage_to` changes). That might not happen very often, if at all, but I think it's a good idea to test (and possibly modify) Phyx.js releases when new versions of these ontologies are released.

Part of the problem here is that we can't version these ontologies semantically, so we can't distinguish between breaking changes and non-breaking changes -- I don't think there's any way we can say that generated phyloreferences should use "the latest Phyloref ontology with minor revisions to version 2018-12-14, but not major revisions". In the absence of that, I don't like the idea that phyloreferences previously generated by Phyx.js might start resolving in different ways if the ontologies change, which is why I like versioned imports. But I'm happy to be overruled here.

hlapp replied:
I see and agree with your point.

hlapp commented on src/wrappers/PhyxWrapper.js:
```diff
@@ -270,12 +270,19 @@ class PhyxWrapper {
+    if (!Array.isArray(jsonld['owl:imports'])) jsonld['owl:imports'] = [jsonld['owl:imports']];
+    jsonld['owl:imports'].push('http://raw.githubusercontent.com/phyloref/curation-workflow/develop/ontologies/phyloref_testcase.owl');
+    jsonld['owl:imports'].push('http://ontology.phyloref.org/2018-12-14/phyloref.owl');
+    jsonld['owl:imports'].push('http://ontology.phyloref.org/2018-12-14/tcan.owl');
```
Same as above.

hlapp commented on src/wrappers/PhyxWrapper.js:
```diff
@@ -270,12 +270,19 @@ class PhyxWrapper {
+    // Set up the ontology imports. If one is present, we add our imports to that.
+    if (!has(jsonld, 'owl:imports')) jsonld['owl:imports'] = [];
+    if (!Array.isArray(jsonld['owl:imports'])) jsonld['owl:imports'] = [jsonld['owl:imports']];
+    jsonld['owl:imports'].push('http://raw.githubusercontent.com/phyloref/curation-workflow/develop/ontologies/phyloref_testcase.owl');
```
If that's indeed the only URL where this is being served, we need to fix this ASAP. Can you file an issue?

gaurav replied:
Actually, we don't need the Testcase Ontology for reasoning any more! I've started work on a PR that removes it entirely (PR #81). We should still move the RDF properties in the `testcase:` namespace that we use into another namespace IMO, which is covered by #78.

hlapp commented on test/examples/correct/testudinata_phylonym.json:
```diff
@@ -0,0 +1,125 @@
+{
+    "@context": "../../../docs/context/development/phyx.json",
```
Remind me why we are using the development version here?

gaurav replied:
The purpose of these example files and the test suite is to test the development version of the Phyx context file -- the code that generates the phyloreferences and the context file need to be tied very closely together during development, so that changes in property names or how they are used are interpreted correctly using the context file to generate valid OWL. The previous version of the Phyx context file will not work on currently generated phyloreferences, because it is out of date in terms of both property names and IRI mappings.

Leaving it pointing to `development` also means that we don't need to change the example files every time we release a new Phyx context file: instead, we can just copy the development file (and the JSON schema) into a new directory (e.g. `v0.3.0` or `v1.0.0`), and then continue development in the `development` directory.

hlapp commented on test/scripts/phyx2owl.js:
```diff
@@ -40,14 +40,14 @@ describe(PHYX2OWL_JS, function () {
     const EXPECTED_OWL_FILE = path.resolve(__dirname, '../examples/correct/brochu_2003.jsonld');
 
-    // If there is already a '../examples/brochu_2003.json' file, we should delete it.
+    // If there is already a '../examples/brochu_2003.owl' file, we should delete it.
```
I assume this is because it gets generated by the tests, right? If so, it seems to be this should be done in a teardown and/or setup script, not somewhere obscure.

gaurav replied:
Yup! I've opened an issue to figure this out (https://github.com/phyloref/phyx.js/issues/82). I've been trying to work around an issue with the JPhyloRef JAR download -- that's something that would also be nice to move into a setup block somewhere.

hlapp reviewed (changes requested):
So some issues pointed out earlier are starting to re-appear, which concerns me. Are we rushing and losing attention to detail?

gaurav commented:
> So some issues pointed out earlier are starting to re-appear, which concerns me. Are we rushing and losing attention to detail?

I hope not, although I took a few days away from the project to clear my head just in case. To my mind, most of our earlier discussion was to do with how the apomorphy was being represented in Phyx and OWL/JSON-LD, which you didn't comment on in this set of feedback. So hopefully that is decided for now! Most of the issues you point to are either related to how we document these new fields or on other details relating to OWL imports and test suite structure, all of which were very useful. I think once we figure out the JSON Schema definition of `phenotypeBearer`, we should be done here? If not, no worries -- we can keep working on this until it's good to go.

hlapp reviewed (approved):
Finally 👍🏻
